### ISSUE TITLE
Persist workflow data in table instead of a JSON field

### ISSUE THREAD

**Issue by acailly** [danielgerlag/workflow-core#377]
We have several workflows to rewrite and we are trying to use this workflow engine to do so

The workflow definition and execution works very well and the flexibility offered by the fluent API is a perfect fit for us

However, another service used to do some extracting and reporting based on data of the workflow instances, which were in dedicated tables and not in a JSON column as it is with the SQL Server persistence provider

Since this service is not able to correctly handle JSON data (due to tools limitation) by themself, we wondered how to make the persistence store the workflow data in a dedicated table instead of a JSON column in the 'Workflow' table

Has anyone tried to do something like that ? or maybe has some advice 😃 ?

I will try to report here my findings

**Comment by danielgerlag** [comment]
You can wire it up to Elasticsearch - https://github.com/danielgerlag/workflow-core/tree/master/src/providers/WorkflowCore.Providers.Elasticsearch
For there you can query Elasticsearch with your custom data object.

**Comment by acailly** [comment]
Yes, I saw the SearchIndex option.

However:
- The other services tools only works in a SQL Server environment (I can't change that)
- The team sometime needs to edit the workflow data, so a query-only database does not fit

**Comment by danielgerlag** [comment]
Honestly, if you could switch to MongoDB, it would be ideal for you as the workflow data is persisted within the Mongo document and is query-able and index-able, even if different workflow types have different kinds of objects for their custom data.  That is the main challenge with a relational DB, is every workflow type would need it's own table.

**Comment by acailly** [comment]
I know, I would have to create a specific db context to handle each workflow type.

But I'm locked in with SQL server and Mongo is not an option.

**Comment by OAlbrecht2407** [comment]
Do you need to access all the properties of the workflow-instance or only the data property of an workflow-instance? 
For the latter i've written my own provider (targeting SAP SQL Anywhere) which saves the data to another table and restores it on reading. In my case it saves the data as json but it's up to you how to save the data as long as your are able to restore ist.

**Comment by acailly** [comment]
That is what I'm trying to do 👍
Actualy, storing the data in JSON in another table seems to be the easy part.
Since I want to store the data the "relational" way (with a schema, not in JSON), I tried to use generics in the provider and making it work with EFCore.
That's the hard part 🙂

**Comment by acailly** [comment]
I managed to do what I wanted to do, unfortunately that means forking the initial project 😢 

I started with changing the type of PersistedWorkflow.Data from object to a generic named `TPersistedWorkflowData`

I made this type implement a new interface named `IPersistedWorkflowData` Because I wanted this type to have some specific fields: PersistenceId, WorkflowId. 

This interface also contains two methods, `fromWorkflowData` and `toWorkflowData`, that embed the conversion logic between Workflow.Data (object) and PersistedWorkflow.Data (TPersistedWorkflowData)

By introducting a generic type in PersistedWorkflow, some other types had to be made generic:

- `PersistedExecutionPointers<TPersistedWorkflowData>`
- `PersistedExecutionPointerCollection<TPersistedWorkflowData>`
- `PersistedExtensionAttribute<TPersistedWorkflowData>`
- `EntityFrameworkPersistenceProvider<TPersistedWorkflowData>`
 
To avoid making WorkflowDbContext, I added an abstract method `Type GetWorkflowDataType()` which return the TPersistedWorkflowData type

```
var workflowDataType = GetWorkflowDataType(); 

var workflowType = typeof(PersistedWorkflow<>).MakeGenericType(workflowDataType);
var workflows = modelBuilder.Entity(workflowType);

[...]

var workflowData = modelBuilder.Entity(workflowDataType);

[...]

var executionPointerType = typeof(PersistedExecutionPointer<>).MakeGenericType(workflowDataType);
var executionPointers = modelBuilder.Entity(executionPointerType);

[...]

var extensionAttributeType = typeof(PersistedExtensionAttribute<>).MakeGenericType(workflowDataType);
var extensionAttributes = modelBuilder.Entity(extensionAttributeType);

[...]

ConfigureWorkflowDataStorage(workflowData); // Added this method too
```

I also tried to make the migration work in the SQL Server persistence, without success.
So I set the `canMigrateDb` always false.

So, finally I can do that:

```
services.AddWorkflow(x => x.UseSqlServer("my connection string", typeof(MyPersistedWorkflowData), true));
```

And the result is that the data is no longer stored in the Workflow table but in the WorkflowData instead

I close this issue but if someone has a better way to do it, I'm interested 😉
